Edit multiple issues, PRs in parallel

heaths opened the pull request:
Resolves #5782 by allowing multiple issues or PRs to be edited in parallel, and querying for shared fields once to reduce network requests.

heaths committed:
Edit multiple issues, PRs in parallel

Resolves #5782 by allowing multiple issues or PRs to be edited in parallel, and querying for shared fields once to reduce network requests.

heaths commented:
This is not complete, but I was hoping for an early review to see if you're good with this approach. It's similar to some other CLI commands and what I've done in some GH extensions.

It allows interactive with only 1 issue per discussion in #5782. Since I check early as part of argument validation, the `editRun` function is roughly the same, with a couple of interactive checks to prompt. That would only actually happen if there was 1 issue, so the logic is left as-is and simple.

There's still a few things to do:

- ~Do the same for `gh pr edit`~ (_will do this in a separate PR_)
- [x] Reorder or refactor how shared editable fields are fetched since the defaults are no longer set. In manual testing, I found that this breaks the flow e.g., selecting "Assignees" does not show anyone or allow you to specify someone and instead just asked to submit, prompting for `Y/n`.
- [x] Add tests for issue's and PR's `editRun` functions.

heaths commented:
@mislav @samcoe what do you think of the direction here? I'll make similar changes to `pr edit` and adds tests for multiple issues for both.

I refactored the brunt of the lookup code into separate methods because so many commands call the singular version, so it seemed like a good idea to consolidate all the checks therein; though, part of me feels that may be unnecessary in an EXE project because those commands only work with a single issue anyway. Perhaps a separate method could just `panic` on completely unexpected results e.g., multiple outputs, which shouldn't happen. Sort of like runtime asserts you should never hit anyway. Thoughts?

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -24,8 +26,9 @@ type EditOptions struct {
-	Interactive bool
+	SelectorArgs []string
+	Interactive  bool
+	Workers      int
```
I don't think we should allow the user to specify the number of workers. Seems like too low level of a detail for them to be concerned with.

heaths replied:
They can't. This is for tests to change it to 1, which is used elsewhere in the CLI. It's not exposed as a flag. I will make it private, though.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,7 +149,15 @@ func editRun(opts *EditOptions) error {
+		}
+	}
+
 	lookupFields := []string{"id", "number", "title", "body", "url"}
```
Seems like these `if` conditions can now remove the `opts.Interactive` condition as we are taking care of that in the survey above.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -157,59 +173,98 @@ func editRun(opts *EditOptions) error {
-		}
+	// Update all issues in parallel.
+	if opts.Workers == 0 {
+		opts.Workers = 10
```
As this is an IO bound operation do we need to limit the workers? Seems likely that the user will run into API limits here before they ever reach a Go routine limit.

heaths replied:
10 workers is used elsewhere throughout the CLI. I figured it was because most services limit the number of concurrent connections. Typically it's 2, and networking stacks will often queue requests but those can timeout. Is that not possible in the CLI?

heaths replied:
Though, seems in most places a worker count is used more where go routines are pulling off a channel like a queue. Still, to make this work for testing, the worker count has to be 1. If left as the zero value, though, I could skip the call to `SetLimit`.

samcoe commented on pkg/cmd/issue/shared/lookup.go:
```diff
@@ -19,26 +19,61 @@ import (
-		}
+// IssuesFromArgWithFields loads 1 oe more issues or pull requests with the specified fields. If some of the fields
+// could not be fetched by GraphQL, this returns non-nil issues and a *PartialLoadError.
+func IssuesFromArgsWithFields(httpClient *http.Client, baseRepoFn func() (ghrepo.Interface, error), args []string, fields []string) ([]*api.Issue, ghrepo.Interface, error) {
```
This feels a bit backwards with `IssueFromArgWithFields` calling `IssuesFromArgWithFields`. It feels like it should be revered where `IssuesFromArgWithFields` is just a loop calling `IssueFromArgWithFields` with some concurrency built in.

heaths replied:
Hadn't considered that. Does seem more elegant.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -113,6 +117,10 @@ func NewCmdEdit(f *cmdutil.Factory, runF func(*EditOptions) error) *cobra.Comman
+			if opts.Interactive && len(opts.SelectorArgs) > 1 {
+				return cmdutil.FlagErrorf("multiple issues cannot be edited interactively")
+			}
+
```
I wonder if this is the spot we should validate that if multiple issues are being edited that they all belong to the same host and same repo. I feel like doing input validation in `RunE` is a pattern we have used to great effect and it would be better to do the validations here before doing any actual work below in `editRun`.

heaths replied:
I could look into it, but it would move parsing the repo from `editRun` to here. Since we were already parsing the repo for each issue - since we have to parse each issue for a possible URL anyway - it made sense to alter that function.

samcoe replied:
That is fair. Let's leave it for now and we can come back to it later if anyone feels strongly enough that the validations should occur here instead.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -157,59 +173,98 @@ func editRun(opts *EditOptions) error {
+	for _, issue := range issues {
+		issue := *issue
+		g.Go(func() error {
+			editable.Title.Default = issue.Title
+			editable.Body.Default = issue.Body
+			editable.Assignees.Default = issue.Assignees.Logins()
+			editable.Labels.Default = issue.Labels.Names()
+			editable.Projects.Default = append(issue.ProjectCards.ProjectNames(), issue.ProjectItems.ProjectTitles()...)
+			projectItems := map[string]string{}
+			for _, n := range issue.ProjectItems.Nodes {
+				projectItems[n.Project.ID] = n.ID
+			}
+			editable.Projects.ProjectItems = projectItems
+			if issue.Milestone != nil {
+				editable.Milestone.Default = issue.Milestone.Title
+			}
```
Using `editable` across the Go routines seems like an issue... I would think it would be safer to create a new `prShared.Editable` for each issue that needs editing.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -157,59 +173,98 @@ func editRun(opts *EditOptions) error {
-	opts.IO.StartProgressIndicator()
-	err = prShared.UpdateIssue(httpClient, repo, issue.ID, issue.IsPullRequest(), editable)
+			// Allow interactive prompts for one issue; failed earlier if multiple issues specified.
+			if opts.Interactive {
+				editorCommand, err := opts.DetermineEditor()
+				if err != nil {
+					return err
+				}
+				err = opts.EditFieldsSurvey(&editable, editorCommand)
+				if err != nil {
+					return err
+				}
+			}
```
It feels odd allowing interactivity in the non-main Go routine. I don't know if it is an issue though. Just thought I would mention it incase that is a rule we shouldn't be breaking.

heaths replied:
Given the validation in `NewCmdEdit`, this can only happen when there's a single issue and stdout/stdin can be safely accessed across go routines or even threads. Given that, it seemed more disruptive to bifurcate interactive and non-interactive modes. This should be safe, but we could consider other options necessary. Resolving this just to keep tracking of pending issues, but feel free to reopen if you want to explore other options.

samcoe replied:
I am okay with keeping it for now. I do think that bifurcating the interactive and non-interactive code paths could lead to more readable and maintainable code though, despite likely having some code duplication.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -157,59 +173,98 @@ func editRun(opts *EditOptions) error {
+
+	issueURLs := make([]string, 0, len(issues))
+	issuesChan := make(chan string)
+	doneCh := make(chan any)
+	go func() {
```
Perhaps I am missing something but could this be simplified to not be run in a Go routine? It would make it so we don't need the extra `doneCh`.

heaths replied:
I didn't have this originally and didn't figure I needed it, but there was a race condition on my laptop where all the issue URLs were sent to the channel, but without signaling this `doneCh` they weren't all received when ranging over the `issueURLs` channel. I've used this pattern before and honestly couldn't figure out why, but having a separate channel like this certainly solves the problem. Any thoughts?

samcoe replied:
I didn't see the code you had originally but if the `issuesChan` was unbuffered then I can see how the code would deadlock on `g.Wait()` and also `issuesChan <- issue.URL`. If you haven't already could you try making `issueChan` with the length of the `issues` slice?

samcoe reviewed (commented):
@heaths Thanks for getting start on this. Left a handful of inline code comments. Let me know if you have any questions about them.

heaths committed:
Resolve PR feedback and add tests

heaths committed:
Clone Editable, simply Go routines

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +150,125 @@ func editRun(opts *EditOptions) error {
+	for _, issue := range issues {
+		// Copy variables to capture in the go routine below.
+		issue := *issue
+		editable := editable
```
As the `prSahred.Editable` struct contains slices I think we are going to have to do a more in depth copy in order to not have data race conditions.

heaths replied:
I added an `Editable.Clone` with documentation as to what it does. I didn't want to call this `Copy` since that might indicate a deep copy, but most fields therein don't need to be copied.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +150,125 @@ func editRun(opts *EditOptions) error {
+				}
+			}
+
+			err = prShared.UpdateIssue(httpClient, repo, issue.ID, issue.IsPullRequest(), editable)
```
What are your thoughts on having the Go routine only contain this function call and moving all the manipulation of `editable` struct and prompting code outside of it? Seems to me that if we can reduce the amount of code that gets executed inside the Go routine to only the operations we need to be async we will reduce potential bugs that could occur.

heaths replied:
The `Editable` will need to be copied, and if we do the interface prompt outside it means bifurcating the code for a single issue vs. multiple. I think it'd make the code more confusing, personally. That's why I added the comment to explain, though, that we could only be interactive with a single issue so there should be no fear of trying to prompt under different go routines.

samcoe replied:
I am not sure I am following the reasoning here. Perhaps I didn't explain my idea well enough though. Was just thinking moving line 205 `g.Go(func() error {` down to line 231. I tried it out locally and it seems to work fine without any additional changes.

samcoe replied:
One added benefit of this is we could allow interactive editing with multiple issues, something that we might want to support in the future.

heaths replied:
I get it now. Thanks for clarifying, and I agree that's a nice benefit if you wanted to support it in the future. Obviously the total execution time is gated on user input, but there'd be no reason not to do the update of the previous issue while prompting for subsequent issues.

In fact, should I just add support for that now. When @mislav and I were discussing this idea in #5782 I don't think we considered that - probably just assumed there'd be a contention issue once parallelized, but with your idea it wouldn't be. Just need to be clear which issue we're prompting for which either 1) changes the UX slightly for a single issue or, 2) we just don't do for a single issue.

samcoe replied:
I think we should hold off on allowing interactivity for the moment since that was not what was initially agreed upon and I would like to run the idea by the team first. I think it would be fairly easy to iterate and add in the feature in a follow up PR if we decide we want it.

heaths replied:
I opened #7285 to track. I'd be happy to implement it when we complete this PR. I'm working on all the changes discussed herein and will apply the same for PRs.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +150,125 @@ func editRun(opts *EditOptions) error {
+	opts.IO.StartProgressIndicatorWithLabel(fmt.Sprintf("Updating %d issues", len(issues)))
+	for _, issue := range issues {
+		// Copy variables to capture in the go routine below.
+		issue := *issue
```
I don't see an issue with this code but I tend to like this pattern a bit more as it seems more clear to me
```go
for i, _ := range issues {
    issue := issues[i]
    ...
}
```

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +149,107 @@ func editRun(opts *EditOptions) error {
+	err = g.Wait()
+	close(issueURLs)
+
 	if err != nil {
```
As we now potentially have some issues being successfully edited and some not I think we alert the user which ones have been successful and which ones either failed or were not attempted. Right now if there is an error the user will just see the error and won't know if any of the issues were properly edited.

heaths replied:
Good point. I should still probably dump those that were successfully edited to stdout. `errgroup` will return on the first error, though because none of the API calls take a `context` I can't pass in a `context` that `errgroup.WithContext` would return, thus cancelling the rest. Thus, I wonder if this err check should just be a return of the `err` or at least moved to the bottom of the function.

Since any existing callers should expect either a single line to stdout (by default) or an exit code, though, do you anticipate any problems? Callers would have to pass in multiple issues, so any existing callers wouldn't do that. If they did, they should expect this new potential mix of stdout + exit code. Seems reasonable, but wanted your take.

samcoe replied:
Yeah... this is getting a bit more complicated unfortunately due to the fact that the network calls do not take a `context`. I think we might need to rethink the strategy a bit. 

I am thinking we move away from using `sync.ErrGroup` and just go with `sync.WaitGroup` here. We will kick off updating each issue and allow each update to run to completion. We can keep track of the ones that were successful and the ones that failed. Then for the output we can indicate which ones successfully updated and which ones failed. If any of them failed then we can return a non-zero exit code.

I think that as long as we keep the same output for when there is only one issue specified then this won't be a breaking change. When a user specifies multiple issues to update then I think it is reasonable that they would expect multiple line output.

Let me know what you think.

heaths replied:
I think `sync.WaitGroup` makes sense but agree it's a good idea to consider passing a `context` to network calls. If/when that happens, this could be updated to an `errgroup.Group`. Worth a `TODO` comment? I'll leave it off for now because I suspect not.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +149,107 @@ func editRun(opts *EditOptions) error {
 
-	fmt.Fprintln(opts.IO.Out, issue.URL)
+	for issueURL := range issueURLs {
+		fmt.Fprintln(opts.IO.Out, issueURL)
```
If we sort the issueURLs before they are output we can control the order of the output for testing and we can do away with setting the number of workers.

samcoe replied:
I am of the mind that we want the tests to actually execute the functionality with parallelization so that we can discover any data race bugs that might be occurring that we wouldn't see by limiting to 1 worker.

heaths replied:
Fair. I was considering sorting originally anyway, but wasn't sure it should be a lexical sort or order based on input. I'll go lexical for now unless you prefer specified ordered.

heaths replied:
To note, until recently when I had to dig more into the request mocks I thought they were actually ordered too, since they tended to be registered in order (which makes sense to "visualize" at least).

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +149,107 @@ func editRun(opts *EditOptions) error {
+			return nil
+		})
+	}
 	opts.IO.StopProgressIndicator()
```
This stop progress indicator should probably move to after `g.Wait()`.

heaths replied:
D'oh.

samcoe commented on pkg/cmd/pr/shared/editable.go:
```diff
@@ -201,6 +201,57 @@ func (e Editable) MilestoneId() (*string, error) {
+func (es *EditableSlice) clone() EditableSlice {
+	cpy := EditableSlice{
+		// Shallow copies since no mutation.
+		Add:     es.Add,
+		Remove:  es.Remove,
```
I think `Add` and `Remove` can be edited by the survey so they will need to be fully copied.

heaths replied:
Seems that I can also do a shallow copy of ProjectItems below since it's just a mapping used to map issue numbers to associative project IDs.

samcoe commented on pkg/cmd/issue/shared/lookup.go:
```diff
@@ -41,6 +41,38 @@ func IssueFromArgWithFields(httpClient *http.Client, baseRepoFn func() (ghrepo.I
+	issues := make([]*api.Issue, 0, len(args))
+	var issuesRepo ghrepo.Interface
+
+	for _, arg := range args {
```
It would be nice to also make the retrieval of the issues parallelized. What do you think?

heaths replied:
Yeah, but I'll use an `errgroup.Group` here because I wouldn't expect any 1 fetch to fail and continue with the others. Ideally, it'd be great to fetch all issues in one request, but seems there's no REST or GraphQL endpoint capable of doing such. Might be able to use aliasing with GraphQL but it's quite a hack and probably not worth it. I'd be surprised if people are updating enough issues to really make it worth it.

samcoe replied:
That makes sense to me to use the `errgroup.Group` here. Agreed, lets not try to optimize for a single request right now, seems like even if it was possible it would be a pretty big refactor.

heaths committed:
Emit successful and failed URLs

heaths commented:
@samcoe I'm done with runtime changes and manual testing but having problems with writing tests. What's used in cli/cli is rather limited, making it difficult to identify specific requests unlike gock used in cli/go-gh. For example, I need to return an error from an `updateIssue` to test the success + failure collection code. Seems the `Matcher` isn't capable, though I haven't fully explored what all other code is potentially doing here. May need to write my own matcher. Or, what about introducing gock into cli/cli?

Updated: I added a matcher for mutations - could do the same for queries, but didn't need it so it'd get no coverage - in lieu of gock and this allows me to test the failure scenarios for both issue queries and updates.

If you're good with this pattern, I'll replicate the same for PRs. Basically, I sort and output all the successful issue URLs first and then the failed URLs since the CLI would return a error as the final line: it'd be all grouped together.

Alternatively, I considered aggregating all errors into a single one, but the CLI currently targets Go 1.19 and not 1.20 so there's no built-in error wrapping. I'm open to changing that for this PR, though.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +148,125 @@ func editRun(opts *EditOptions) error {
+	failedIssueChan := make(chan string, len(issues))
+	g := sync.WaitGroup{}
+
+	opts.IO.StartProgressIndicatorWithLabel(fmt.Sprintf("Updating %d issues", len(issues)))
```
I think that due to code structure we need this `start` and corresponding `stop` to only occur in non-interactive mode else the `EditFieldsSurvey` will interrupt the progress indicator.

samcoe commented on pkg/cmd/issue/edit/edit.go:
```diff
@@ -141,75 +148,125 @@ func editRun(opts *EditOptions) error {
+	}
+
+	if len(failedIssueErrors) > 0 {
+		return fmt.Errorf("failed to update %d issue(s)", len(failedIssueErrors))
```
We can use `text.Pluralize` to make sure issue has the correct plurality here.

heaths replied:
I considered that, but I thought if anyone was ever checking stdout against some pattern, that would be a bit annoying since the text could change. But if that's not really much a concern, I will use `text.Pluralize`.

samcoe replied:
I think it is okay to use `text.Pluralize`. My thought would be that the user should really be checking the exit code first and then if it is non-zero the user will likely need to investigate the output manually rather than trying to rely on pattern matching the output to determine what failed.

samcoe commented on pkg/cmd/issue/shared/lookup.go:
```diff
@@ -41,6 +43,60 @@ func IssueFromArgWithFields(httpClient *http.Client, baseRepoFn func() (ghrepo.I
+
+	var issuesRepo ghrepo.Interface
+	issuesChan := make(chan *api.Issue, len(args))
+	mu := sync.Mutex{}
```
Having to think about the errgroup and mutex feels a bit complicated to me. What do you think about doing the repo validations before starting the parallel work? An additional benefit of this is that we don't do any extra network calls that might not be needed. Something like this:
```go
// IssuesFromArgWithFields loads 1 or more issues or pull requests with the specified fields.
// If some of the fields could not be fetched by GraphQL, this returns non-nil issues and a *PartialLoadError.
func IssuesFromArgsWithFields(httpClient *http.Client, baseRepoFn func() (ghrepo.Interface, error), args []string, fields []string) ([]*api.Issue, ghrepo.Interface, error) {
	var repo ghrepo.Interface
	var numbers []int

	for _, a := range args {
		number, r, err := IssueNumberAndRepoFromArg(a)
		if err != nil {
			return nil, nil, err
		}
		numbers = append(numbers, number)
		if r == nil {
			continue
		}
		if repo == nil {
			repo = r
			continue
		}
		if !ghrepo.IsSame(repo, r) {
			return nil, nil, fmt.Errorf(
				"multiple issues must be in same repo: found %q, expected %q",
				ghrepo.FullName(repo),
				ghrepo.FullName(r),
			)
		}
	}

	if repo == nil {
		var err error
		if repo, err = baseRepoFn(); err != nil {
			return nil, nil, err
		}
	}

	issuesChan := make(chan *api.Issue, len(args))
	g := errgroup.Group{}
	for _, num := range numbers {
		n := num
		g.Go(func() error {
			issue, err := findIssueOrPR(httpClient, repo, n, fields)
			if err != nil {
				return err
			}

			issuesChan <- issue
			return nil
		})
	}
	err := g.Wait()
	close(issuesChan)

	if err != nil {
		return nil, nil, err
	}

	issues := make([]*api.Issue, 0, len(args))
	for issue := range issuesChan {
		issues = append(issues, issue)
	}

	return issues, repo, nil
}
```

heaths replied:
That's probably better. I was trying to reuse the logic rather than dup it, but the benefits here are worth it.

heaths replied:
I'll need to resolve the `baseRepo` using the `baseRepoFn` earlier or the `IsSame` check will be useless, but this otherwise works.

samcoe commented on pkg/cmd/issue/edit/edit_test.go:
```diff
@@ -351,38 +546,48 @@ func Test_editRun(t *testing.T) {
 			assert.Equal(t, tt.stdout, stdout.String())
-			assert.Equal(t, tt.stderr, stderr.String())
+			// Use regex match since mock errors and service errors will differ.
+			assert.Regexp(t, regexp.MustCompile(tt.stderr), stderr.String())
```
```suggestion
			assert.Regexp(t, tt.stderr, stderr.String())
```
Looks like `assert.Regexp` will take care of the regexp compiling for us.

samcoe commented on pkg/cmd/issue/edit/edit_test.go:
```diff
@@ -311,11 +342,175 @@ func Test_editRun(t *testing.T) {
+		{
+			name: "non-interactive multiple issues with fetch failures",
+			input: &EditOptions{
+				SelectorArgs: []string{"9999"},
```
Let's add another selectorArg so this is actually multiple issues.

samcoe commented on pkg/httpmock/stub.go:
```diff
@@ -57,6 +57,33 @@ func GraphQL(q string) Matcher {
 	}
 }
 
+func MatchGraphQLMutation(q string, cb func(map[string]interface{}) bool) Matcher {
```
I like this. I think naming it `GraphQLMutationMatcher` is slightly better although we don't have a set naming scheme for matchers.

samcoe reviewed (commented):
This is looking real good, almost there I think! Appreciate your patients going through this review process.

heaths committed:
Compare repos outside goroutines

Also resolves other PR feedback

heaths commented:
@samcoe as far as issues go, this should be ready. I rebased on trunk but it may take a little more time to make the changes to PRs. When I first looked at this, I took a cursory glance and it seemed that, as with some other places in the CLI, issues and PRs shared more code than they actual do once I started digging in.

On that note, would you consider taking issues first? And should I refactor PRs or issues to actually share more? Issues have the function we'd be discussing while PRs uses a Finder. PRs also don't prompt for what to edit, but I don't intend to change that unless you wanted me to. I'm trying to keep refactoring to a minimum to reduce risk, so for now I'll continue with such minimal changes unless you want me to unify the code more.

samcoe reviewed (approved):
@heaths This looks great to me! Thanks for the hard work on this. 

I don't see a problem with merging/releasing this before `pr edit` gets changed. I don't think they need to be shipped together, each feature provides value on its own.

For `pr edit` I think keeping the change set to a minimum is the right approach and we shouldn't try to refactor the two commands to share more code unless it really makes sense. As you said, the two commands have some differing behavior so I think it is completely fine if we end up with some duplicate code.

heaths commented:
@samcoe I've got a few ideas of how to support this for PRs while keeping changes to a minimum. Almost all PR commands use the `Finder` and `Finder.Find` would be a lot to duplicate as I did with `IssuesFromArgsWithFields`, but like that function simply calling one within another doesn't work ideally. What if, instead, I do what I did originally with `IssuesFromArgsWithFields` and had a singular function that calls the plural variant and either assumes or asserts a single PR is returned (probably safe to assume)? `Finder.Find` is already using `errgroup` so adding another goroutine seems insignificant.

The idea would be that no selectors means the current branch as it is currently, but then condition into a loop of 1 or more PR numbers, URLs, or branch names. `Find` could be a wrapper that assumes a single issue is returned, while `MultiFind` ("multi" seems to be a common enough prefix in the codebase, but I'm open to other suggestions, of course) can return multiple. That was initially the goal of `IssuesFromArgsWithFields` but I appreciate it wasn't using `errgroup` or even a wait group before, but `Find` already does.

If you'd prefer - I know we overlap a bit in your afternoon / my morning - perhaps we could chat over Teams.

samcoe commented:
@heaths Agreed that changing `finder.Find` would cause too much of a wide spreading change and that adding a new function is the the better approach. I am partial to the name `finder.FindMultiple` but I don't have strong opinions on it. This overall approach seems good to me.

heaths commented:
@samcoe I more or less started down that path, though with a different name. I like yours better and will change it.

The only widespread change I ended up doing was making `FindOptions.Selector` and `BaseBranch` `[]string` and renaming them plurally. The single + multiple fields just felt so wrong I couldn't bring myself to do it. Sure, it touched ~10 files but it's all 1- or 2-liners. I'll continue with that for a separate PR.

mislav commented on pkg/cmd/issue/shared/lookup.go:
```diff
@@ -41,6 +36,70 @@ func IssueFromArgWithFields(httpClient *http.Client, baseRepoFn func() (ghrepo.I
 
+// IssuesFromArgWithFields loads 1 or more issues or pull requests with the specified fields. If some of the fields
+// could not be fetched by GraphQL, this returns non-nil issues and a *PartialLoadError.
+func IssuesFromArgsWithFields(httpClient *http.Client, baseRepoFn func() (ghrepo.Interface, error), args []string, fields []string) ([]*api.Issue, ghrepo.Interface, error) {
```
Due to the async nature of this function's implementation, the resulting slice of `api.Issue` results does not necessarily match the order of input `args`. It doesn't seem like that negatively affects the rest of the `edit` implementation, but I wonder will this be apparent to future callers of this function. Just a thought!

heaths replied:
Order shouldn't matter, true, and since it didn't even handle multiple issues before I don't foresee any regression. That said, @samcoe and I did briefly discuss whether to just lexically sort the issues or try to match the order in which they were specified - the latter of which didn't seem necessary. Could still be done, though. I would hope - and see on reason - callers would care about the order specified.

mislav reviewed (approved):
Thanks for the hard work on this @heaths and @samcoe for stewarding this! 🚀
